feat: Simprocs for the Working Mathematician (#97)
leanprover-community/blog: posts/fantastic-simprocs.md

```lean
example : Finset.Icc 1 0 = ∅ := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 1 = {1} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 4 = {1, 2, 3, 4} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc a (a + 2) = {a, a + 1, a + 2} := by
  -- fails: the bounds of the interval aren't numerals!
  simp only [Icc_ofNat_ofNat]
```

#### The `Nat.reduceDvd` simproc

The `Nat.reduceDvd` simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.

```lean
example : 3 ∣ 9 := by
  simp only [Nat.reduceDvd]

example : ¬ 2 ∣ 49 := by
  simp only [Nat.reduceDvd]
  -- Remaining goal: `¬ False`
  simp

example : a ∣ a * b := by
  simp only [Nat.reduceDvd] --fails
```

Here the metaprogram run by `Nat.reduceDvd` does the following whenever an expression of the form `a ∣ b` (where `a, b` are natural numbers) is encountered:
- Check that `∣` is the usual natural number division.
- Check that `a` and `b` are numerals.
- Compute `b % a`.
- If this quantity is zero, then return `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.

> - If this quantity isn't zero, then return `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.

Reviewer: I am not sure how I feel about these technicalish explanations being so early in the blog post. Should they be relegated outside of `## Examples`? to the second blog post? I am open to suggestions

Reviewer 2: I'm pretty content with this as it is: you mention the right hand side is computed from the left, and giving an overview of how that computation could work is helpful imo. I don't see this as too technical (although you probably could omit the part about checking `∣`)

Author: I agree - this is something I meant to remove (e.g. "Check that `∣` is the usual natural number division." is a very confusing sentence if you can't see the corresponding code)

> - If this quantity isn't zero, then return `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.

Reviewer 2: At this point in the story, it's not clear to the reader what's meant by "return". I think it would be helpful to give a rough intuition of how a simproc works, before the examples start, saying something like "the simproc's job is to try to find an appropriate RHS, and if so, return it together with a proof that LHS = RHS. Then `simp` will perform this replacement, using the returned proof term as appropriate." (Adjust phrasing to your taste!)

  -- Works since `simp` can simplify `37 * 0` to `True`
  -- because it knows the lemma `mul_zero a : a * 0 = 0`.
  simp only [↓reduceIte]

example (X : Type) (P : Prop) (a b : X) : (if P ∨ ¬ P then a else b) = a := by
  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
  simp only [↓reduceIte]

open Classical

example : (if FermatLastTheorem then 1 else 2) = 1 := by
  --This fails because `simp` can't simplify `FermatLastTheorem` to `True` or `False`
  simp only [↓reduceIte]
  -- Remaining goal: `⊢ (if FermatLastTheorem then 1 else 2) = 1`
  sorry -- See https://imperialcollegelondon.github.io/FLT for how to solve this `sorry` ;)
```

The down arrow `↓` in `simp only [↓reduceIte]` indicates that `reduceIte` is being called as a *preprocedure*.
This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.
Concretely, `simp` needs only simplify one of the expressions `a` and `b`, as the other one is discarded upfront.
For example, `↓reduceIte` allows `simp` to avoid touching `bigComplicatedExpression` at all in the following:
```
example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
  simp only [↓reduceIte]
```

> Notice that one could always replace `simp [↓reduceIte]` with `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter would simplify both branches of the `ite`, instead of merely the one that is actually relevant.

Reviewer 2: I misunderstood this at first read. What about:
```suggestion
Notice that one could always replace `simp [↓reduceIte]` with `simp [ite_cond_eq_true, ite_cond_eq_false]` which would leave the same final goal state, but the would simplify inside both branches of the `ite`, instead of simplifying away the irrelevant one before passing to subexpressions.
```

---
author: 'Yaël Dillies, Paul Lezeau'
category: 'Metaprogramming'
date: 2025-03-29 12:33:00 UTC+00:00
description: 'Brief introduction to simprocs and what they are useful for'
has_math: true
link: ''
slug: fantastic-simprocs
tags: 'simp, simproc, meta'
title: Fantastic Simprocs and Where to Find Them.
type: text
---

Lean core added support for custom simplification procedures, aka *simprocs*.
This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.

<!-- TEASER_END -->

This post describes purely informally what simprocs are and do.
The second post will be a walkthrough to writing a simple simproc in three different ways.

# What is a simproc

To understand what a simproc is and how it works, we will first explain how `simp` works.
Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.

## How simp works

> On an informal level, `simp` can be seen as recursively traversing the expression `e` to be simplified from the outside-in, each time checking whether the expression matches the left hand side `LHS` of some simp lemma of the form `LHS = RHS`, and replacing `LHS` by `RHS` in the expression if it does match.

Reviewer 2: This appears inaccurate: note that the Lean reference manual says this: "The [simp](https://lean-lang.org/doc/reference/latest/Tactic-Proofs/Tactic-Reference/#simp) tactics primarily rewrite from the inside out."

Reviewer: This is the same phrasing issue as here: https://github.com/leanprover-community/blog/pull/97#discussion_r2040673512

namely simp *traverses* from the outside-in, and *rewrites* from the inside-out

Reviewer 2: Right, and I think that since the default and common case is postprocedures, it makes more sense for the informal description to talk about that case instead.

Author: Maybe it's worth mentioning also that any lemma (and simproc) can be set to act during the outside-in part, by adding the `\d` arrow in front of the identifier.

Author: I've also pushed some local changes that I had forgotten to add, which remove the "outside-in" from this sentence and give a tiny bit more information about how simp traverses expressions.

For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
```lean
⊢ 37 * (Nat.fib 0 + 0) = 0
⊢ 37 * (0 + 0) = 0
⊢ 37 * 0 = 0
⊢ 0 = 0
⊢ True
```

>! To see what proof steps `simp` used, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.

In this picture, simp lemmas are *fixed* rules to turn a *specific* left hand side into a *specific* right hand side.
In contrast, simprocs are *flexible* rules to turn a *specific* left hand side into a right hand side *computed* from the left hand side.
In this sense, they are *parametric simp lemmas*.

## Examples of simprocs

In this section, we exemplify three simprocs through the following use cases:
* Computation
* Avoiding combinatorial explosion of lemmas
* Performance optimisation

> ### Computation:

Reviewer 2: In each of the examples in this section, `decide` works to prove the goal, so it's not clear to me that this section works as a good example of why simprocs are useful.

Reviewer: I've added an explanation of why `decide` bad. Hopefully that's enough

> #### The `Finset.Icc_ofNat_ofNAt` simproc

Reviewer 2: ```suggestion
#### The `Finset.Icc_ofNat_ofNat` simproc
```

> The `Finset.Icc_ofNat_ofNAt` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.

Reviewer 2: ```suggestion
The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.
```

This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.
Concretely, `simp` needs only simplify one of the expressions `a` and `b`, as the other one is discarded upfront.
For example, `↓reduceIte` allows `simp` to avoid touching `bigComplicatedExpression` at all in the following:
```
example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
  simp only [↓reduceIte]
```

Notice that one could always replace `simp [↓reduceIte]` with `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter would simplify both branches of the `ite`, instead of merely the one that is actually relevant.

Internally, this simproc is a small metaprogram that does the following whenever an expression of the form `ite P a b` is encountered:
- Call `simp` on `P` to get a simplified expression `P'` and a proof `h` that `P = P'`.
- If `P'` is `True` then return the simplified expression `a` and the proof `ite_cond_eq_true r` that `ite P a b = a`
- If `P'` is `False` then return the simplified expression `b` and the proof `ite_cond_eq_false r` that `ite P a b = b`
- Otherwise, let `simp` continue the search.

### Many more applications!

> At the end of this blog post, we will see how to build step by step a simproc for computing a variant of `List.range` when the parameter is a natural number literal.

Reviewer 2: I presume you want to change this :)

---
author: 'Yaël Dillies, Paul Lezeau'
category: 'Metaprogramming'
date: 2025-03-29 12:33:00 UTC+00:00
description: 'Brief introduction to simprocs and what they are useful for'
has_math: true
link: ''
slug: fantastic-simprocs
tags: 'simp, simproc, meta'
title: Fantastic Simprocs and Where to Find Them.
type: text
---

Lean core added support for custom simplification procedures, aka *simprocs*.
This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.

<!-- TEASER_END -->

This post describes purely informally what simprocs are and do.
The second post will be a walkthrough to writing a simple simproc in three different ways.

# What is a simproc

To understand what a simproc is and how it works, we will first explain how `simp` works.

> Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.

Reviewer 2: Perhaps I'm missing something, but I don't actually see an explanation of what simprocs do here! You explain what simp does, then have a single sentence giving one intuition for what a simproc is, but you don't say what it actually does. And then you jump to the examples. You also mention non-examples and pointers to other theorem provers, but I don't see that in this post either?

Reviewer: There are now several sentences. I'm not necessarily happy with the amount of explanations though

> In particular, by default `simp` only attempts to use lemmas on `e` *on the way out*.

For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
```lean
⊢ 37 * (Nat.fib 0 + 0) = 0
⊢ 37 * (0 + 0) = 0
⊢ 37 * 0 = 0
⊢ 0 = 0
⊢ True
```

> To see what proof steps `simp` used, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.

In this picture, simp lemmas are *fixed* rules to turn a left hand side of a *specific* shape into a right hand side of a *specific* shape.
In contrast, simprocs are *flexible* rules to turn a left hand side of a *specific* shape into a right hand side of a shape *computed* from the left hand side.
In this sense, they are *parametric simp lemmas*.
To make the previous description a little more precise, the way `simp` works is by recursively traversing an expression `e` (and its subexpressions), and trying to match these with the "left hand side" of a simproc.

## Examples of simprocs

> In this section, we exemplify three simprocs through the following use cases:

Reviewer 2: ```suggestion
In this section, we exemplify four simprocs through the following use cases:
```

This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.
Concretely, `simp` needs only simplify one of the expressions `a` and `b`, as the other one is discarded upfront.
For example, `↓reduceIte` allows `simp` to avoid touching `bigComplicatedExpression` at all in the following:
```
example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
  simp only [↓reduceIte]
```

Notice that one could always replace `simp [↓reduceIte]` with `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter would simplify both branches of the `ite`, instead of merely the one that is actually relevant.

Internally, this simproc is a small metaprogram that does the following whenever an expression of the form `ite P a b` is encountered:
- Call `simp` on `P` to get a simplified expression `P'` and a proof `h` that `P = P'`.
- If `P'` is `True` then return the simplified expression `a` and the proof `ite_cond_eq_true r` that `ite P a b = a`
- If `P'` is `False` then return the simplified expression `b` and the proof `ite_cond_eq_false r` that `ite P a b = b`

> - Otherwise, let `simp` continue the search.

Reviewer 2: ```suggestion
```
This is not mentioned in any of the other simproc descriptions!

Then we will give some examples and non-examples of simprocs.

# How simp works

`simp` is made of two components.

The first component is **rewriting rules**.
Almost all rewriting rules are lemmas that prove an equality and are tagged with `@[simp]` in eg Lean or Mathlib.
A rewriting rule is characterised by its *left hand side* and *right hand side*.
Eg for a lemma of the form `LHS = RHS`, this is `LHS` and `RHS` respectively.

The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.

For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
```lean
⊢ 37 * (Nat.fib 0 + 0) = 0
⊢ 37 * (0 + 0) = 0
⊢ 37 * 0 = 0
⊢ 0 = 0
⊢ True
```

> The order in which `simp` traverses an expression is relatively intricate.
> As a simple approximation, simplification is performed from the inside-out.
> See the next blog post for full details on the algorithm.

> > If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.

Reviewer 2: Do you definitely want all of this as two block quotes? I don't think they're significant enough for that!

Reviewer: Oh I thought that was pretty nice. Imagining myself reading this blog post as a beginner, these block quotes are answering the exact two questions I think I would be asking myself.

Or do you mean that there's better formatting than a block quote? I was using that as an aparté

Reviewer 2: I think they can just be text as usual: block quotes in this way are usually used to indicate key takeaways from the overall post; these are useful points but I don't think they need highlighting in this way.

Reviewer 2: But this is predicated on the way block quotes are displayed in the blog; perhaps the disagreement here is just that we're picturing different things!

# What is a simproc?

Earlier, we said "almost all rewriting rules are lemmas".
What about rewriting rules that are not lemmas?
This is where simprocs enter the picture.

A simproc is a program which, given an expression `LHS`, computes a new expression `RHS` and finds a proof of `LHS = RHS` on the fly.

The concept of a simproc is genuinely more powerful than that of a simp lemma.

> Even though the right hand side of a lemma changes depending on its left hand side, it only does so via *syntactic substitution*.

Reviewer 2: ```suggestion
This is because even though the right hand side of a lemma changes depending on its left hand side, it only does so via *syntactic substitution*.
```
or some other phrasing which clarifies that this sentence is an explanation of the previous

For example, the lemma `Nat.mul_add (a b c : Nat) : a * (b + c) = a * b + a * c` can be specialized to have right hand side `0 * b + 0 * c`, `a * b + a * (c + d)`, etc...
But all these have *shape* `_ * _ + _ * _` and eg no value of `a`, `b`, `c` will make it have shape `_ * _ + _ * _ + _ * _`.

With this in mind, simprocs are *parametric simp lemmas*.

## Everything is a simproc

> As a very neat unification of concepts, it turns out that `simp` only ever handles simprocs.

Reviewer 2: ```suggestion
As a very neat unification of concepts, it turns out that the implementation of `simp` only ever handles simprocs.
```

Reviewer: Hmm, your suggestion sounds a bit clunky

Reviewer 2: The distinction I want to draw here is that you first said `simp` is made of two components, then you're saying that actually the first one was a lie. And so adding "implementation" here makes it clearer that these are consistent viewpoints, one is how simp intuitively works and the other is how it's practically implemented. I'm happy with alternate phrasing to a similar effect, but I think the current phrasing gives a confusing message here.

Every simp lemma is turned into a simproc before being fed to `simp`.

# Examples of simprocs

In this section, we exemplify four simprocs through the following use cases:
* Computation
* Avoiding combinatorial explosion of lemmas
* Performance optimisation

## Computation:

### The `Finset.Icc_ofNat_ofNat` simproc

The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.

```lean
example : Finset.Icc 1 0 = ∅ := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 1 = {1} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 4 = {1, 2, 3, 4} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc a (a + 2) = {a, a + 1, a + 2} := by
  -- fails: the bounds of the interval aren't numerals!
  simp only [Icc_ofNat_ofNat]

> ```

Reviewer 2: imo this one is less compelling than the `Nat.reduceDvd` one, and it currently doesn't have the informal description that all the others do. I don't mind keeping it, but I think it should go after the reduceDvd one, and have some explanation of the metaprogram (which can be one/two lines)

```lean
theorem exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ b, p b → a = b) :
    (∃ b, p b) = p a
```

## Performance optimisation: The `reduceIte` simproc

The `reduceIte` simproc is designed to take expressions of the form `if P then a else b` (aka `ite P a b`) and replace them with `a` or `b`, depending on whether `P` simplify to `True` or `False`.

```lean
example : (if 37 * 0 then 1 else 2) = 1 := by
  -- Works since `simp` can simplify `37 * 0` to `True`
  -- because it knows the lemma `mul_zero a : a * 0 = 0`.
  simp only [↓reduceIte]

example (X : Type) (P : Prop) (a b : X) : (if P ∨ ¬ P then a else b) = a := by
  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
  simp only [↓reduceIte]

open Classical

example : (if FermatLastTheorem then 1 else 2) = 1 := by
  --This fails because `simp` can't simplify `FermatLastTheorem` to `True` or `False`
  simp only [↓reduceIte]
  -- Remaining goal: `⊢ (if FermatLastTheorem then 1 else 2) = 1`
  sorry -- See https://imperialcollegelondon.github.io/FLT for how to solve this `sorry` ;)
```

The down arrow `↓` in `simp only [↓reduceIte]` indicates that `reduceIte` is being called as a *preprocedure*.

> This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.

Reviewer 2: You could mention here that this is an example of why "simplification is performed from the inside-out" was a simplification (no pun intended), which a) gives a tiny bit more context on what it actually happening and b) makes it clearer that you're not self-contradicting

---
author: 'Yaël Dillies, Paul Lezeau'
category: 'Metaprogramming'
date: 2025-04-16 12:00:00 UTC+01:00
description: 'Informal introduction to simprocs and what they are useful for'
has_math: true
link: ''
slug: fantastic-simprocs
tags: 'simp, simproc, meta'
title: Fantastic Simprocs and Where to Find Them.
type: text
---

> Lean core added support for custom simplification procedures, aka *simprocs*.

Reviewer 3: ```suggestion
Lean v4.6.0 added support for custom simplification procedures, aka *simprocs*.
```
perhaps with some remark about this being a long time ago!

> This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.

Reviewer 3: Style question; should we write `` `simproc` `` everywhere?

Reviewer: ... or `Simproc`? I have been following the convention that formal concepts are quoted, while informal ones aren't. In this blog post, a simproc is an informal concept (but `simp` is a formal one), and simprocs become the formal `Simproc`s in the followup blog post

Reviewer 3: Both `Lean.Meta.Simp.Simproc` and `simproc` are valid pieces of lean syntax, so I think the latter is fine to use instead of the former.

My heuristic is "if it's not a real word then always write and quote it as code", but it's not my blog post so I'll let you make the final call.

Reviewer 2: For comparison, the reference manual also refers to simproc without code indicators, so this is at least consistent with that

Reviewer 3: Can we link the reference manual from the blog post?

Reviewer 2: Sure, I think it makes sense to mention it at the end of the "how simp works" section, saying something like, you can read about it in more detail here: https://lean-lang.org/doc/reference/latest////The-Simplifier/#the-simplifier

Reviewer 2: There's not much about simprocs there, only a paragraph here https://lean-lang.org/doc/reference/latest////The-Simplifier/Rewrite-Rules/#simp-rewrites

Reviewer 2: You could also link https://leanprover-community.github.io/extras/simp.html#introduction

Reviewer 4: My vote is to italicize it on first use but otherwise treat it like a regular word (which seems to be already happening).

<!-- TEASER_END -->

This post describes purely informally what simprocs are and do.
The second post will be a walkthrough to writing a simple simproc in three different ways.

To understand what a simproc is and how it works, we will first explain how `simp` works.
Then we will give some examples and non-examples of simprocs.

# How simp works

`simp` is made of two components.

The first component is **rewriting rules**.

> Almost all rewriting rules are lemmas that prove an equality and are tagged with `@[simp]` in eg Lean or Mathlib.

Reviewer 3: ```suggestion
Almost all rewriting rules are lemmas that prove an equality or `↔` and are tagged with `@[simp]` in eg Lean or Mathlib.
```

A rewriting rule is characterised by its *left hand side* and *right hand side*.
Eg for a lemma of the form `LHS = RHS`, this is `LHS` and `RHS` respectively.

The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.

For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
```lean
⊢ 37 * (Nat.fib 0 + 0) = 0
⊢ 37 * (0 + 0) = 0
⊢ 37 * 0 = 0
⊢ 0 = 0
⊢ True
```

> The order in which `simp` traverses an expression is relatively intricate.
> As a simple approximation, simplification is performed from the inside-out.
> See the next blog post for full details on the algorithm.

> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.

# What is a simproc?

Earlier, we said "almost all rewriting rules are lemmas".
What about rewriting rules that are not lemmas?

> This is where simprocs enter the picture.

Reviewer 3: Where do theorems like `: Even 0` fit into this narrative? Perhaps worth a footnote that theorem of the form `P` are often treated as `P ↔ True`?

Reviewer: I've added an explanation earlier

But all these have *shape* `_ * _ + _ * _` and eg no value of `a`, `b`, `c` will make it have shape `_ * _ + _ * _ + _ * _`.

With this in mind, simprocs are *modular simp lemmas*.

## Everything is a simproc

As a very neat unification of concepts, it turns out that `simp`'s implementation only ever handles simprocs.
Every simp lemma is turned into a simproc before being fed to `simp`.

# Examples of simprocs

In this section, we exemplify four simprocs through the following use cases:
* Avoiding combinatorial explosion of lemmas
* Computation
* Performance optimisation

## Avoiding combinatorial explosion of lemmas: The `existsAndEq` simproc

The `existsAndEq` simproc is designed to simplify expressions of the form `∃ x, ... ∧ x = a ∧ ...` where `a` is some quantity independent of `x` by removing the existential quantifier and replacing all occurences of `x` by `a`.

```lean
example (p : Nat → Prop) : ∃ x : Nat, p x ∧ x = 5 := by
  simp only [existsAndEq]
  -- Remaining goal: `⊢ p 5`

example (p q : Nat → Prop) : ∃ x : Nat, p x ∧ x = 5 ∧ q x := by
  simp only [existsAndEq]
  -- Remaining goal: `⊢ p 5 ∧ q 5`
```

> When presented with a left hand side of the form `∃ x, P x, where `P x` is of the form `_ ∧ ... ∧ _`, `existsAndEq` does the following:

Reviewer 3: ```suggestion
When presented with a left hand side of the form `∃ x, P x`, where `P x` is of the form `_ ∧ ... ∧ _`, `existsAndEq` does the following:
```

- Recursively traverse `P x` inside the existential quantifier looking for an equality `x = a` for some `a`.
- If an equality is found, construct a proof that `∀ x, p x → x = a`.

> - Return the right hand side `P a` together with the proof obtained from the following lemma:
> ```lean
> lemma exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ x, p x → x = a) :
>     (∃ x, p x) = p a
> ```

Reviewer 3: ````suggestion
- Return the right hand side `P a` together with the proof obtained from the following lemma:
  ```lean
  lemma exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ x, p x → x = a) :
      (∃ x, p x) = p a
  ```
````

> Warning: The `existsAndEq` simproc is under active development and the underlying algorithm might change in the future.
> The above description is only accurate at the time of writing (April 2025).

## Computation

Computations are an integral part of theorem proving, and as such there are many ways to perform them.
For example, you will find that the `decide` tactic closes all the examples in this subsection.
There are a few reasons why simprocs are interesting for computation regardless:
* **`decide` relies on decidability instances**.
  Not everything one may want to compute is decidable, and not every decidability instance is efficient.
  In fact, most `Decidable` instances in Lean and Mathlib are very generic, and therefore unspecific and inefficient.

>   Using a simproc gives the opportunity to use a domain-specific algorithm, which is more likely to be efficient and does not rely the decidability machinery.

Reviewer 2: ```suggestion
  Using a simproc gives the opportunity to use a domain-specific algorithm, which is more likely to be efficient and does not rely on the decidability machinery.
```

When presented with a left hand side of the form `ite P a b`, `reduceIte` does the following:
- Call `simp` on `P` to get a simplified expression `P'` and a proof `h` that `P = P'`.
- If `P'` is `True` then return the right hand side `a` together with the proof `ite_cond_eq_true r` that `ite P a b = a`.
- If `P'` is `False` then return the the right hand side `b` togethr with the proof `ite_cond_eq_false r` that `ite P a b = b`.

## Many more applications!

In the second blog post, we will see how to build step by step a simproc for computing a variant of `List.range` when the parameter is a numeral.

# A few caveats

The current design of simprocs comes with a few restrictions that are worth keeping in mind:
* By definition, **a simproc can only be used in `simp`** (and `simp`-like tactics like `simp_rw`, `simpa`, `aesop`), even though the notion of a "modular lemma" could be useful in other rewriting tactics like `rw`.
* **One cannot provide arguments to a simproc to restrict the occurrences it rewrites**.

>   In contrast, this is possible for lemmas in all rewriting tactics: eg `rw [add_comm c]` turns `⊢ a + b = c + d` into `⊢ a + b = d + c` where `rw [add_comm]` would instead have turned it into `⊢ b + a = c + d`.

Reviewer 3: I was hoping this would work, but alas:
```lean
def myProc (a : Nat) : Lean.Meta.Simp.Simproc := fun e => do
  return .continue

-- illegal
simproc_pattern% _ => myProc 

-- ok, but arguments were consumed already
def wat := myProc 1
simproc_pattern% _ => wat
```

Reviewer: Yes sadly I think a few things to change for Lean to acquire `rwproc`

Then we will give some examples and non-examples of simprocs.

# How simp works

`simp` is made of two components.

The first component is **rewriting rules**.
Almost all rewriting rules are lemmas that prove an equality `=` or iff `↔ ` and are tagged with `@[simp]` in eg Lean or Mathlib.
A rewriting rule is characterised by its *left hand side* and *right hand side*.
Eg for a lemma of the form `LHS = RHS` or `LHS ↔ RHS`, this is `LHS` and `RHS` respectively.
If a lemma proves `P` that is not of the form `_ = _` or `_ ↔ _`, it is turned into `P = True`.

The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.

For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
```lean
⊢ 37 * (Nat.fib 0 + 0) = 0
⊢ 37 * (0 + 0) = 0
⊢ 37 * 0 = 0
⊢ 0 = 0
⊢ True
```

> The order in which `simp` traverses an expression is relatively intricate.
> As a simple approximation, simplification is performed from the inside-out.

> > See the next blog post for full details on the algorithm.

Reviewer 5: ```suggestion
> See the next blog post for more details on the algorithm.
```
unless you really mean *full*!

> > If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.

Reviewer 5: ```suggestion
> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `MyGoal`.
```

* **`decide` cannot compute the right hand side**, given the left hand side only.
  `decide` only works on goals that do not contain any metavariable.
  This rules out using `decide` to discover out what a left hand side is equal to.
  One would need to write down the right hand side we are looking for in order for `decide` to show that it's equal to the left hand side.

### The `Nat.reduceDvd` simproc

The `Nat.reduceDvd` simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.

```lean
example : 3 ∣ 9 := by
  simp only [Nat.reduceDvd]

example : ¬ 2 ∣ 49 := by
  simp only [Nat.reduceDvd]
  -- Remaining goal: `¬ False`
  simp

example : a ∣ a * b := by
  simp only [Nat.reduceDvd] --fails
```

When presented with a left hand side of the form `a ∣ b` where `a` and `b` are natural numbers, `Nat.reduceDvd` does the following:
- Check that `a` and `b` are numerals.
- Compute `b % a`.
- If `b % a` is zero, then return the right hand side `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.

> - If `b % a` isn't zero, then return the right hand side `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.

Reviewer 2: In the ite example you give the proof as well as saying what it's a proof of, I think it's clearer to do the same here too:
```suggestion
- If `b % a` isn't zero, then return the right hand side `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl` that `b % a = 0 = False`.
```

Reviewer 2: (and similarly for the `True` case)

### The `Finset.Icc_ofNat_ofNat` simproc

> The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.

Reviewer 2: It might be worth giving a link to docs for `Finset.Icc`; I'd imagine some readers might not know what it is!

```lean
example : Finset.Icc 1 0 = ∅ := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 1 = {1} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 4 = {1, 2, 3, 4} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc a (a + 2) = {a, a + 1, a + 2} := by
  -- fails: the bounds of the interval aren't numerals!
  simp only [Icc_ofNat_ofNat]
```

When presented with a left hand side of the form `Finset.Icc a b` where `a` and `b` are natural numbers, `Finset.Icc_ofNat_ofNat` does the following:
* Check that `a` and `b` are numerals.
* compute the expression `{a, ..., b}` recursively on `b`, along with the proof that it equals `Finset.Icc a b`.

## Performance optimisation: The `reduceIte` simproc

> The `reduceIte` simproc is designed to take expressions of the form `if P then a else b` (aka `ite P a b`) and replace them with `a` or `b`, depending on whether `P` simplify to `True` or `False`.

Reviewer: [The Cambridge dictionary lists "aka"](https://dictionary.cambridge.org/dictionary/english/aka#google_vignette)

---
author: 'Yaël Dillies, Paul Lezeau'
category: 'Metaprogramming'
date: 2025-04-16 12:00:00 UTC+01:00
description: 'Informal introduction to simprocs and what they are useful for'
has_math: true
link: ''
slug: fantastic-simprocs
tags: 'simp, simproc, meta'

> title: Fantastic Simprocs and Where to Find Them.

Reviewer 6: I know the title is a slight joke, but I would actually be curious to know if there is a way to find or search for simprocs in mathlib? E.g., if I'm in the middle of a proof, is there some way for me to check if there is a simproc that would be helpful here? (Right now it seems to me like I have to keep track of what simprocs mathlib has and what their names are).

Reviewer 3: ```lean
run_meta do
  let s ← Lean.Meta.Simp.getSimprocs
  Lean.logInfo m!"{s.pre.toArray ++ s.post.toArray}"
  pure ()
```

Reviewer 3: I think that only finds `simproc`s (the ones enabled automatically), not `simproc_decls` (the ones you refer to by name). Loogle should be able to find both.

Author: It would be pretty cool to have a command ``#simprocs` that takes in a pattern and outputs the list of `simproc`s that activate on that pattern, e.g something like `#simprocs (ite _ _ _)` would find `iteReduce`

Author: I've changed the title to _Simprocs for the Working (Formal) Mathematician_, which I think make the more practical purpose of this post clearer (and we can repurpose the previous name for the last post in the series where we actually explain how to write a simproc)

type: text
---

Lean v4.6.0 (back in February 2024!) added support for custom simplification procedures, aka *simprocs*.
This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.

<!-- TEASER_END -->

This post describes purely informally what simprocs are and do.
The second post will be a walkthrough to writing a simple simproc in three different ways.

To understand what a simproc is and how it works, we will first explain how `simp` works.
Then we will give some examples and non-examples of simprocs.

# How `simp` works

`simp` is made of two components.

The first component is **rewriting rules**.
Almost all rewriting rules are lemmas that prove an equality `=` or iff `↔` and are tagged with `@[simp]` in e.g. Lean or Mathlib.
A rewriting rule is characterised by its *left hand side* and *right hand side*.

> Eg for a lemma of the form `LHS = RHS` or `LHS ↔ RHS`, this is `LHS` and `RHS` respectively.

Reviewer: If it reads patronising, then you haven't understood it! The point is that `LHS = RHS` is a very special case of a rewriting rule, and the general case is a simproc.

I don't think I can remove the explanation entirely either since I refer to the "left hand side" and "right hand side" of a rewriting rule several times below, in a precise sense.

I've tried something, but I'm not sure it's an improvement

⊢ 37 * 0 = 0
⊢ 0 = 0
⊢ True
```

> The order in which `simp` traverses an expression is relatively intricate.
> As a simple approximation, simplification is performed from the inside-out.
> See the next blog post for more details on the algorithm.

> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `MyGoal`.

# What is a simproc?

Earlier, we said "almost all rewriting rules are lemmas".
What about rewriting rules that are not lemmas?
This is where simprocs enter the picture.

A simproc is a program which, given an expression `LHS`, computes a new expression `RHS` and finds a proof of `LHS = RHS` on the fly.

The concept of a simproc is genuinely more powerful than that of a simp lemma.
This is because the right hand side of a lemma changes only via *syntactic substitution*.

For example, the lemma `Nat.mul_add (a b c : Nat) : a * (b + c) = a * b + a * c` can be specialized to have right hand side `0 * b + 0 * c`, `a * b + a * (c + d)`, etc...
But all these have *shape* `_ * _ + _ * _` and e.g. no value of `a`, `b`, `c` will make it have shape `_ * _ + _ * _ + _ * _`.

> 

Reviewer: The point I'm trying to make here is the distinction that Bhavik was explaining here: https://github.com/leanprover-community/blog/pull/97#discussion_r2021952082

> 

Reviewer 2: My reading of Kevin's message is that the point isn't clearly made in this paragraph: if a reader is confused by this they're not going to come to this PR and go and find my comment!

> 

Author: I think this paragraph has now been removed!

With this in mind, simprocs are *modular simp lemmas*.

## Everything is a simproc

As a very neat unification of concepts, it turns out that `simp`'s implementation only ever handles simprocs.
Every simp lemma is turned into a simproc before being fed to `simp`.

# Examples of simprocs

In this section, we exemplify four simprocs through the following use cases:
* Avoiding combinatorial explosion of lemmas
* Computation
* Performance optimisation

## Avoiding combinatorial explosion of lemmas: The `existsAndEq` simproc

The [`existsAndEq`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=existsAndEq#doc) simproc is designed to simplify expressions of the form `∃ x, ... ∧ x = a ∧ ...` where `a` is some quantity independent of `x` by removing the existential quantifier and replacing all occurences of `x` by `a`.

> 

Reviewer: Because it might move! The `find` page is what the Zulip linkifier uses, and you don't see people complaining that links in old posts are dead, so this is most definitely the right move.

  lemma exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ x, p x → x = a) :
      (∃ x, p x) = p a
  ```

> Warning: The `existsAndEq` simproc is under active development and the underlying algorithm might change in the future.
> The above description is only accurate at the time of writing (April 2025).

## Computation

Computations are an integral part of theorem proving, and as such there are many ways to perform them.
For example, you will find that the `decide` tactic closes all the examples in this subsection.
There are a few reasons why simprocs are interesting for computation regardless:
* **`decide` relies on decidability instances**.
  Not everything one may want to compute is decidable, and not every decidability instance is efficient.
  In fact, most `Decidable` instances in Lean and Mathlib are very generic, and therefore unspecific and inefficient.
  Using a simproc gives the opportunity to use a domain-specific algorithm, which is more likely to be efficient and does not rely on the decidability machinery.
* **`decide` cannot compute the right hand side**, given the left hand side only.
  `decide` only works on goals that do not contain any metavariable.

>   This rules out using `decide` to discover out what a left hand side is equal to.

Reviewer 5: ```suggestion
  This rules out using `decide` to discover what a left hand side is equal to.
```
or maybe "find out"?

  One would need to write down the right hand side we are looking for in order for `decide` to show that it's equal to the left hand side.

### The `Nat.reduceDvd` simproc

> The [`Nat.reduceDvd`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=Nat.reduceDvd#doc) simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.

Reviewer: Thanks, I thought I had already replaced "literal" with "numeral" everywhere but apparently not

```lean
example : 3 ∣ 9 := by
  simp only [Nat.reduceDvd]

example : ¬ 2 ∣ 49 := by
  simp only [Nat.reduceDvd]
  -- Remaining goal: `¬ False`
  simp

example : a ∣ a * b := by
  simp only [Nat.reduceDvd] --fails
```

When presented with a left hand side of the form `a ∣ b` where `a` and `b` are natural numbers, `Nat.reduceDvd` does the following:

> - Check that `a` and `b` are numerals.

Reviewer: Indeed :)

- Compute `b % a`.

> - If `b % a` is zero, then return the right hand side `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl` that `b % a = 0 = True`.
> - If `b % a` isn't zero, then return the right hand side `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl` that `b % a = 0 = False`.

Reviewer 7: ```suggestion
- If `b % a` is zero, then return the right hand side `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl` that `(b % a = 0) = True`.
- If `b % a` isn't zero, then return the right hand side `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl` that `(b % a = 0) = False`.
```

### The `Finset.Icc_ofNat_ofNat` simproc

`Finset.Icc a b` for `a` and `b` in a partial order is the finite set of elements lying between `a` and `b`.

> The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form [`Finset.Icc a b`](https://leanprover-community.github.io/mathlib4_docs/search.html?q=Finset.Icc) where `a` and `b` are numerals, and simplify them to an explicit set.

Reviewer 7: Should this not link to the actual docs page instead? Or is that being less stable a problem?

Reviewer 3: This should link to the magic links that Zulip uses.

```lean
example : Finset.Icc 1 0 = ∅ := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 1 = {1} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc 1 4 = {1, 2, 3, 4} := by
  simp only [Icc_ofNat_ofNat]

example : Finset.Icc a (a + 2) = {a, a + 1, a + 2} := by
  -- fails: the bounds of the interval aren't numerals!
  simp only [Icc_ofNat_ofNat]
```

When presented with a left hand side of the form `Finset.Icc a b` where `a` and `b` are natural numbers, `Finset.Icc_ofNat_ofNat` does the following:
* Check that `a` and `b` are numerals.
* compute the expression `{a, ..., b}` recursively on `b`, along with the proof that it equals `Finset.Icc a b`.

## Performance optimisation: The `reduceIte` simproc

> The `reduceIte` simproc is designed to take expressions of the form `if P then a else b` (aka `ite P a b`) and replace them with `a` or `b`, depending on whether `P` simplify to `True` or `False`.

Reviewer 7: ```suggestion
The `reduceIte` simproc is designed to take expressions of the form `if P then a else b` (aka `ite P a b`) and replace them with `a` or `b`, depending on whether `P` simplifies to `True` or `False`.
```

```lean

> example : (if 37 * 0 then 1 else 2) = 1 := by

Reviewer 5: ```suggestion
example : (if 37 * 0 = 0 then 1 else 2) = 1 := by
```
this is missing right?  :smile:

>   -- Works since `simp` can simplify `37 * 0` to `True`

Reviewer 5: ```suggestion
  -- Works since `simp` can simplify `37 * 0 = 0` to `True`
```

  -- because it knows the lemma `mul_zero a : a * 0 = 0`.
  simp only [↓reduceIte]

example (X : Type) (P : Prop) (a b : X) : (if P ∨ ¬ P then a else b) = a := by
  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
  simp only [↓reduceIte]

> open Classical

Reviewer: We can remove it once you are done proving `FermatLastTheorem` 😉

example : (if FermatLastTheorem then 1 else 2) = 1 := by
  --This fails because `simp` can't simplify `FermatLastTheorem` to `True` or `False`
  simp only [↓reduceIte]
  -- Remaining goal: `⊢ (if FermatLastTheorem then 1 else 2) = 1`
  sorry -- See https://imperialcollegelondon.github.io/FLT for how to solve this `sorry` ;)
```

The down arrow `↓` in `simp only [↓reduceIte]` indicates that `reduceIte` is being called as a *preprocedure*.
This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.

> Recall that we said that, as a simple approximation, simplification is performed from the inside-out.

> > This is a concrete example where simplification happens outside-in.

Reviewer: We're simplifying `ite P a b` before `a` and `b`. This is what's "outside-in". It is true that in order to simplify `ite P a b` we also simplify `P`, but that's not what this remark is about.

> > This is a concrete example where simplification happens outside-in.

Reviewer: I've clarified, hopefully it's better now

Concretely, `simp` needs only simplify one of the expressions `a` and `b`, as the other one is discarded upfront.
For example, `↓reduceIte` allows `simp` to avoid touching `bigComplicatedExpression` at all in the following:
```
example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
  simp only [↓reduceIte]
```

> Notice that `simp [↓reduceIte]` is equivalent to `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter simplifies inside both branches of the `ite`, instead of simplifying away the irrelevant one before passing to subexpressions.

Reviewer 5: ```suggestion
Notice that `simp [↓reduceIte]` is ultimately equivalent to `simp [ite_cond_eq_true, ite_cond_eq_false]`, but the latter simplifies inside both branches of the `ite`, instead of simplifying away the irrelevant one before passing to subexpressions.
```
since otherwise "equivalent" could be interpreted as "behaves the same", but then you explain that it does not!

When presented with a left hand side of the form `ite P a b`, `reduceIte` does the following:
- Call `simp` on `P` to get a simplified expression `P'` and a proof `h` that `P = P'`.
- If `P'` is `True` then return the right hand side `a` together with the proof `ite_cond_eq_true r` that `ite P a b = a`.

> - If `P'` is `False` then return the right hand side `b` together with the proof `ite_cond_eq_false r` that `ite P a b = b`.

Reviewer 8: I have now read and understood all of this example and am completely confused about why we need a `simproc` to do this. If the *only* reason is "saves us from simplifying b if P is true and b is complicated" then I think this example should either be deleted, or rewritten to explain that this is the point. I am not entirely clear on why `simp` can't solve this without a simproc so my comments on this example should be taken with a pinch of salt, but if I'm right and the sole point is avoiding simplifying a complicated expression b then I think this example should *start* with: "Consider the following goal:"
```
example : (if 37 * 0 = 0 then 0 else bigComplicatedExpression) = 0 := by
```
and then explain the problem with running `simp`, and then explain the simproc, and then explain the down-arrow, and then write
```
  simp only [↓reduceIte]
```
and now everything is clear. 

Indeed this works fine:
```
example : 0 = if True then 0 else 1 := by
  simp [-reduceIte]
```
so it seems to me that the only point of this simproc is to save you from simplifying the irrelevant possible answer. You should say this right at the start.

## Many more applications!

In the second blog post, we will see how to build step by step a simproc for computing a variant of `List.range` when the parameter is a numeral.

# A few caveats

The current design of simprocs comes with a few restrictions that are worth keeping in mind:
* By definition, **a simproc can only be used in `simp`** (and `simp`-like tactics like `simp_rw`, `simpa`, `aesop`), even though the notion of a "modular lemma" could be useful in other rewriting tactics like `rw`.
* **One cannot provide arguments to a simproc to restrict the occurrences it rewrites**.

>   In contrast, this is possible for lemmas in all rewriting tactics: e.g. `rw [add_comm c]` turns `⊢ a + b = c + d` into `⊢ a + b = d + c` where `rw [add_comm]` would instead have turned it into `⊢ b + a = c + d`.

Reviewer 7: Is this true? Can you not do `P : Nat -> Simproc` and `simp [P 37]`?

Reviewer 3: [no](https://github.com/leanprover-community/blog/pull/97#discussion_r2045348464)

Reviewer: As explained in the other blog post (#98), simp doesn't consume bare elements of type `Simproc`, but instead elements of type `Simproc` annotated with extra data, like whether it's pre or post. So `simp [(something : Simproc)]` doesn't even make sense

Reviewer: That too. "It wasn't intended" is probably the answer

Reviewer 2: I think it's a bit more than just the parser: there's also a check to make sure what you passed in was `Simproc` (and not, eg `_ -> Simproc`), and the simprocs are stored in an array which is also expecting stuff of that type. So "it wasn't intended" sounds closer to the truth

> * By definition, **a simproc can only be used in `simp`** (and `simp`-like tactics like `simp_rw`, `simpa`, `aesop`), even though the notion of a "modular lemma" could be useful in other rewriting tactics like `rw`.

Reviewer 8: Why mention `aesop`? Why not also mention `norm_num`, that runs `simp`. You could say "`simp`** (and tactics which call `simp` such as `simpa`, `aesop` etc)

# Examples of simprocs

> In this section, we exemplify four simprocs that cover the following following use cases:

Reviewer 3: I'm not following the following following following. I think it should be removed.

example : 3 ∣ 9 := by
  simp only [Nat.reduceDvd]

-- `decide` cannot close this goal.

> example (f : Prop → Prop) : f (2 ∣ 4) = f True := by

Reviewer 2: This is a nice example!
